lib/bnchmrkr.rb

```diff
@@ -1,100 +1,101 @@
 #!/usr/bin/ruby
+# Bnchmrkr is a tool to help Benchmark.measure {} and compare different method implementations
+
+$LOAD_PATH << sprintf('%s/../lib', File.dirname(__FILE__))
+require 'bnchmrkr/mark'
 
 require 'benchmark'
 
-# Bnchmrkr is a tool to help Benchmark.measure {} and compare different method implementations
 class Bnchmrkr
 
-  attr_reader :count, :results
-
-  def initialize(lambdas, count = 100)
-
-    @count    = count
-    @results  = Hash.new
-    @lambdas  = lambdas
+  attr_reader :executions, :marks, :fastest, :slowest
 
-    @fastest = { :name => :unknown, :measure => 2 ** 10 } # TODO really need to find a better max int
-    @slowest = { :name => :unknown, :measure => 0 }
+  def initialize(lambdas, executions = 100)
+    @executions  = executions
+    @marks       = Hash.new
 
-    @cache = Hash.new
+    @fastest = nil
+    @slowest = nil
 
-    @lambdas.each_pair do |name, l|
+    lambdas.each_pair do |name, l|
       unless name.class.eql?(Symbol) and l.class.eql?(Proc)
         raise ArgumentError.new(sprintf('expecting[Symbol,Proc], got[%s,%s]', name.class, l.class))
       end
+
+      @marks[name] = Bnchmrkr::Mark.new(name, l)
     end
 
-    raise ArgumentError.new(sprintf('expecting[Fixnum], got[%s]', count.class)) unless count.class.eql?(Fixnum)
+    raise ArgumentError.new(sprintf('expecting[Fixnum], got[%s]', executions.class)) unless executions.class.eql?(Fixnum)
 
   end
 
-  # return list of named lambdas known
+  # return list of named Bnchmrkr::Marks
   def types
-    @lambdas.keys
+    @marks.keys
   end
 
   # 10 lines to actually do the work..
   def benchmark!
-    @lambdas.each_pair do |name, l|
-      1.upto(@count).each do |round|
-        begin
-          measure = Benchmark.measure {
-            l.call
-          }
-          add_measure(name, measure)
-        rescue => e
-          add_measure(sprintf('%s-failed', name).to_sym, Benchmark.measure {})
-        end
+    @marks.each_pair do |_name, mark|
+      1.upto(@executions).each do |_execution|
+        measure = Benchmark.measure { mark.lambda.call }
+        mark.add_measure(measure)
       end
+
+      mark.compute # this is a safer place to do it than by computing on each measure, but still should consider putting this behind a flag
     end
 
+    calculate_overall
   end
 
-  ## :specific => fastest, slowest, mean, median, total per lambda
-  ## :overall  => fastest, slowest for all
   def inspect
-    { :overall  => calculate_overall, :specific => calculate_per_lambda }
+    return {:foo => { :bar => 'baz'}} if @fastest.nil?
```

Comment:
don't actually care, only a problem when RubyMine tries to .to_s the object before it's created